### ISSUE TITLE
PanToNode with Cola Force Layout not working

### ISSUE THREAD

**Issue by nicorivat** [swimlane/ngx-graph#210]
**Describe the bug**
Hi,

I'm having an issue when I try to pan to a specific node.

![Capture d’écran 2019-06-10 à 19 24 00](https://user-images.githubusercontent.com/15010744/59214219-ef809180-8bb6-11e9-9379-bf39c6ab12a4.png)

I'm using a custom Cola Force layout (custom because I removed the nodes animations when expanding). But when I try to pan to this node:

Before Pan:
![Capture d’écran 2019-06-10 à 19 24 00 copie](https://user-images.githubusercontent.com/15010744/59214303-23f44d80-8bb7-11e9-948d-59944c8630f9.png)

After Pan:
![Capture d’écran 2019-06-10 à 19 24 39](https://user-images.githubusercontent.com/15010744/59214320-2b1b5b80-8bb7-11e9-9291-37ac361c8722.png)

The view is absolutely not centered to the node.

Are you aware of this issue ? Am I doing something wrong ? Tell me if you need other informations.

Thank you

**Expected behavior**
The view must be centered to the node.

**ngx-graph version**
Angular 8.0.0 - Ngx-Graph 6.0.1

**Comment by marjan-georgiev** [comment]
Pan to node should be working on all layouts. We will take a look, thanks for reporting!

**Comment by nicorivat** [comment]
Any news about this issue ?

**Comment by Fahnenschwinger** [comment]
Hey, 

I am having somewhat the same issue although I am using a cluster layout.
When I want to pan to a node the whole graph just seems to move to the bottom/right corner of the viewport regardless of the node, which I want to pan to.

Not sure if I did something wrong, help would be greatly appreciated!

**Comment by Fahnenschwinger** [comment]
Yeah, there is definitely something wrong with some of the Layouts.
The normal dagre Layout works fine. The others e.g dagreCluster etc. don't work like expected.

The problem seems to be that the nodes all have the position x, y of 0 which they shouldn't have.
And then when the panTo method is used
```
const panX = -this.panOffsetX - x * this.zoomLevel + this.dims.width / 2;
                   = 0         = 0       = 0        = half to the right
 
const panY = -this.panOffsetY - y * this.zoomLevel + this.dims.height / 2;
                   = 0         = 0       = 0        = half to the bottom
```

There is a problem with this part of the code which sets the positions:
```
const dagreToOutput = node => {
      const dagreNode = this.dagreGraph._nodes[node.id];
      return {
        ...node,
        position: {
          x: dagreNode.x,
          y: dagreNode.y
        },
        dimension: {
          width: dagreNode.width,
          height: dagreNode.height
        }
      };
    };
    graph.clusters = (graph.clusters || []).map(dagreToOutput);
    graph.nodes = graph.nodes.map(dagreToOutput);
```

The graph.nodes array has the right values.
But the dagreGraph._nodes array has all positions set to 0

I am not really familiar with these expressions so I don't really understand what's happening here.

But the dagre Layout has a different approach which seems to set the Positions correctly

```
 for (const dagreNodeId in this.dagreGraph._nodes) {
      const dagreNode = this.dagreGraph._nodes[dagreNodeId];
      const node = graph.nodes.find(n => n.id === dagreNode.id);
      node.position = {
        x: dagreNode.x,
        y: dagreNode.y
      };
      node.dimension = {
        width: dagreNode.width,
        height: dagreNode.height
      };
    }
```

**Comment by Fahnenschwinger** [comment]
I fixed it. 
```
run(graph: Graph): Graph {
    this.createDagreGraph(graph);
    dagre.layout(this.dagreGraph);
    graph.edgeLabels = this.dagreGraph._edgeLabels;
    
    for (const dagreNodeId in this.dagreGraph._nodes) {
      const dagreNode = this.dagreGraph._nodes[dagreNodeId];
      const ngxElements = graph.nodes.concat(graph.clusters);
      const node = ngxElements.find(o => o.id === dagreNode.id);
    
      node.position = {
        x: dagreNode.x,
        y: dagreNode.y
      };
      node.dimension = {
        width: dagreNode.width,
        height: dagreNode.height
      };
    }

    return graph;
}
```

This worked for the clusterLayout.

I am not quite sure what's causing the problem in the first Place
but the positions are set to 0 after the dagreToOutput variable
and after they get mapped they are correct again.

**Comment by vladimir-spintech** [comment]
I was able to work around this bug for Cola Force Layout by importing this library module directly to my app and then modifying [this line](https://github.com/swimlane/ngx-graph/blob/c86f0c881b1c01cd8fd433010935e8e06022c776/projects/swimlane/ngx-graph/src/lib/graph/graph.component.ts#L1027):
`const node = this.nodes.find(n => n.id === nodeId);` to 
`const node = this.graph.nodes.find(n => n.id === nodeId);`

I'm not sure if this is an ultimate fix but it looks like the position of nodes in the original `this.nodes` array is never updated.

**Comment by assafsun** [comment]
Is this still relevant? I think the problem is inside the pan to node function where we should find the node from this.graph.nodes and not this.nodes

**Comment by vladimir-spintech** [comment]
hi @assafsun,

Yes that's my understanding of the problem as well. I guess this is still relevant because pan to node function is broken in master.

**Comment by assafsun** [comment]
@vladimir-spintech , I read now your comment and saw that you figure it out so good catch 🙂

I created a PR fix for this issue.

**Comment by assafsun** [comment]
Fixed and in version 7.0.0
